File: evaluation-and-observability/3. Python/observability.py
# Observability and Evaluation Custom Solution for Amazon Bedrock Applications
import pytz
import json
import time
import boto3
from uuid import uuid4
from datetime import datetime, timezone
from typing import Any, Callable, Dict, Optional

class BedrockLogs:
    VALID_FEATURE_NAMES = ["None", "Agent", "KB", "InvokeModel"]

    def __init__(self, delivery_stream_name: str = None, 
                 experiment_id: str = None, 
                 default_call_type: str = 'LLM', 
                 feature_name: str = None, 
                 feedback_variables: bool = False
                ):
        self.delivery_stream_name = delivery_stream_name
        self.experiment_id = experiment_id
        self.default_call_type = default_call_type
        self.feedback_variables = feedback_variables

        if feature_name is not None:
            if feature_name not in BedrockLogs.VALID_FEATURE_NAMES:
                raise ValueError(f"Invalid feature_name '{feature_name}'. Valid values are: {', '.join(BedrockLogs.VALID_FEATURE_NAMES)}")
        self.feature_name = feature_name
        self.step_counter = 0

        if self.delivery_stream_name is None:
            raise ValueError("delivery_stream_name must be provided or set equals to 'local' example: delivery_stream_name='local'.")

        if self.delivery_stream_name == 'local':
            self.firehose_client = None
        else:
            self.firehose_client = boto3.client('firehose')

    @staticmethod
    def find_keys(dictionary, key, path=[]):
        """
        Recursive function to find all keys in a nested dictionary and their paths.

        Args:
            dictionary (dict): The dictionary to search.
            key (str): The key to search for.
            path (list, optional): The path of keys to the current dictionary. Defaults to None.

        Returns:
            list: A list of tuples containing the key's path and value.
        """
        results = []

        if isinstance(dictionary, dict):
            for k, v in dictionary.items():
                new_path = path + [k]
                if k == key:
                    results.append((new_path, v))
                else:
                    results.extend(BedrockLogs.find_keys(v, key, new_path))
        elif isinstance(dictionary, list):
            for i, item in enumerate(dictionary):
                new_path = path + [i]
                results.extend(BedrockLogs.find_keys(item, key, new_path))

        return results

    def extract_session_id(self, log_data: Dict[str, Any]) -> str:
        """
        Extracts the session ID from the log data. If the session ID is not available,
        it generates a new UUID for the run ID.

        Args:
            log_data (Dict[str, Any]): The log data dictionary.

        Returns:
            str: The session ID or a newly generated UUID if the session ID is not available.
        """
        if self.feature_name == "Agent":
            session_id_paths = self.find_keys(log_data, 'x-amz-bedrock-agent-session-id')
        else:
            session_id_paths = self.find_keys(log_data, 'sessionId')

        if session_id_paths:
            path, session_id = session_id_paths[0]
            return session_id
        else:
            return str(uuid4())

    def handle_agent_feature(self, output_data, request_start_time):
        """
        Handles the logic for the 'Agent' feature, including step counting and latency calculation.

        Args:
            output_data (Any): The output data from the function call.
            request_start_time (float): The start time of the request.

        Returns:
            Any: The updated output data with step numbers and latency information.
        """
        self.session_id = None
        prev_trace_time = None
        for data in output_data:
            if isinstance(data, dict) and 'trace' in data:
                trace = data['trace']
                if 'start_trace_time' in trace:
                    # Check if 'start_trace_time' is defined correctly
                    if not isinstance(trace['start_trace_time'], float):
                        raise ValueError("The key 'start_trace_time' should be present and should be a time.time() object.")

                    # Calculate the latency between traces
                    if prev_trace_time is None:
                        trace['latency'] = trace['start_trace_time'] - request_start_time
                    else:
                        trace['latency'] = trace['start_trace_time'] - prev_trace_time

                    prev_trace_time = trace['start_trace_time']
                    trace['step_number'] = self.step_counter
                    self.step_counter += 1
                    data['trace'] = trace  # Update the 'trace' dictionary in the original data

            elif isinstance(data, list):
                for item in data:
                    if isinstance(item, dict) and 'start_trace_time' in item:
                        # Check if 'start_trace_time' is defined correctly
                        if not isinstance(item['start_trace_time'], float):
                            raise ValueError("The key 'start_trace_time' should be present and should be a time.time() object.")

                        # Calculate the latency between traces
                        if prev_trace_time is None:
                            item['latency'] = item['start_trace_time'] - request_start_time
                        else:
                            item['latency'] = item['start_trace_time'] - prev_trace_time

                        prev_trace_time = item['start_trace_time']
                        item['step_number'] = self.step_counter
                        self.step_counter += 1

                    elif isinstance(item, dict) and 'trace' in item:
                        trace = item['trace']
                        if 'start_trace_time' in trace:
                            # Check if 'start_trace_time' is defined correctly
                            if not isinstance(trace['start_trace_time'], float):
                                raise ValueError("The key 'start_trace_time' should be present and should be a time.time() object.")

                            # Calculate the latency between traces
                            if prev_trace_time is None:
                                trace['latency'] = trace['start_trace_time'] - request_start_time
                            else:
                                trace['latency'] = trace['start_trace_time'] - prev_trace_time

                            prev_trace_time = trace['start_trace_time']
                            trace['step_number'] = self.step_counter
                            self.step_counter += 1
                            item['trace'] = trace  # Update the 'trace' dictionary in the original item

        return output_data

    def watch(self, capture_input: bool = True, capture_output: bool = True, call_type: Optional[str] = None):
        def wrapper(func):
            def inner(*args, **kwargs):
                # For Latency Calculation:
                self.request_start_time = time.time()

                # Get the function name
                function_name = func.__name__

                # Capture input if requested
                input_data = args if capture_input else None
                input_log = None
                if input_data:
                    input_log = input_data[0]
                    
                # Generate observation_id
                observation_id = str(uuid4())
                obs_timestamp = datetime.now(timezone.utc).isoformat()

                # Get the start time
                start_time = time.time()

                # Calls the function to be executed
                result = func(*args, **kwargs)
                
                # Capture output if requested
                output_data = result if capture_output else None

                # Get the end time
                end_time = time.time()

                # Calculate the duration
                duration = end_time - start_time
                
                # Begin Logging Time:
                logging_start_time = time.time()

                # Handle the 'Agent' feature case
                if self.feature_name == "Agent":
                    if output_data is not None:
                        output_data = self.handle_agent_feature(output_data, self.request_start_time)
                        run_id = self.extract_session_id(output_data[0])
                    else:
                        run_id = self.extract_session_id(input_log)
                else:
                    # Extract the session ID from the log or generate a new one
                    run_id = self.extract_session_id(input_log)

                # Prepare the metadata
                metadata = {
                    'experiment_id': self.experiment_id,
                    'run_id': run_id,
                    'observation_id': observation_id,
                    'obs_timestamp': obs_timestamp,
                    'start_time': datetime.fromtimestamp(start_time, tz=pytz.utc).isoformat(),
                    'end_time': datetime.fromtimestamp(end_time, tz=pytz.utc).isoformat(),
                    'duration': duration,
                    'input_log': input_log,
                    'output_log': output_data,
                    'call_type': call_type or self.default_call_type,
                    'feature_name': self.feature_name,
                    'feedback_enabled': self.feedback_variables
                }

                # Update the metadata with additional_metadata if provided
                additional_metadata = kwargs.get('additional_metadata', {})
                if additional_metadata:
                    metadata.update(additional_metadata)

                input_data = kwargs.get('user_prompt', {})
                if input_data:
                    metadata.update(input_data)
                    
                # Get the end time
                logging_end_time = time.time()

                # Calculate the duration
                logging_duration = logging_end_time - logging_start_time
                metadata['logging_duration'] = logging_duration

                # Send the metadata to Amazon Kinesis Data Firehose or return it locally for testing:
                if self.delivery_stream_name == 'local':
                    if self.feedback_variables:
                        print("Logs in local mode-with feedback:")
                        return result, metadata, run_id, observation_id
                    else:
                        print("Logs in local mode-without feedback:")
                        return result, metadata
                # log to firehose
                else:
                    firehose_response = self.firehose_client.put_record(
                        DeliveryStreamName=self.delivery_stream_name,
                        Record={
                            'Data': json.dumps(metadata)
                        }
                    )
                    if self.feedback_variables:
                        print("Logs in S3-with feedback:")
                        return result, run_id, observation_id
                    else:
                        print("Logs in S3-without feedback:")
                        return result

            return inner
        return wrapper

File: evaluation-and-observability/2. JavaScript/observability.mjs
import { v4 as uuidv4 } from 'uuid';
import AWS from 'aws-sdk';
import {performance} from 'perf_hooks';


/* Defines the class that sets the context for the observability conditions including 
   what feature of Bedrock to observe and log data to return */

class BedrockLogs {
    
    /* Defines the class that sets the context for the observability conditions including 
   what feature of Bedrock to observe and log data to return */
   
    static VALID_FEATURE_NAMES = ["None", "Agent", "KB", "InvokeModel"];

    constructor(deliveryStreamName = null, experimentId = null,defaultCallType = 'LLM', featureName = null,feedbackVariables=false) {
        this.deliveryStreamName = deliveryStreamName;
        this.experimentId = experimentId;
        this.defaultCallType = defaultCallType;
        this.feedbackVariables = feedbackVariables;
        if (featureName !== null) {
            if (!BedrockLogs.VALID_FEATURE_NAMES.includes(featureName)) {
                throw new Error(`Invalid feature_name '${featureName}'. Valid values are: ${BedrockLogs.VALID_FEATURE_NAMES.join(', ')}`);
            }
        }
        this.featureName = featureName;
        this.stepCounter = 0;

        if (this.deliveryStreamName === null) {
            throw new Error("delivery_stream_name must be provided or set equals to 'local' example: delivery_stream_name='local'.");
        }

        if (this.deliveryStreamName === 'local') {
             this.firehoseClient=null;
        }
        else{
             this.firehoseClient = new AWS.Firehose();
        }

    }

    static findKeys(dictionary, key, path = []) {
        const results = [];

        if (typeof dictionary === 'object' && dictionary !== null) {
            if (Array.isArray(dictionary)) {
                dictionary.forEach((item, i) => {
                    const newPath = [...path, i];
                    results.push(...BedrockLogs.findKeys(item, key, newPath));
                });
            } else {
                Object.entries(dictionary).forEach(([k, v]) => {
                    const newPath = [...path, k];
                    if (k === key) {
                        results.push([newPath, v]);
                    } else {
                        results.push(...BedrockLogs.findKeys(v, key, newPath));
                    }
                });
            }
        }

        return results;
    }

    extractSessionId(logData) {
        
        /* Extracts the session ID from the log data. If the session ID is not available,
        it generates a new UUID for the run ID.*/

        let sessionIdPaths;
        if (this.featureName === "Agent") {
            sessionIdPaths = BedrockLogs.findKeys(logData, 'sessionId');
        } else {
            sessionIdPaths = BedrockLogs.findKeys(logData, 'sessionId');
        }

        if (sessionIdPaths.length > 0) {
            const [path, sessionId] = sessionIdPaths[0];
            return sessionId;
        } else {
            return uuidv4();
        }
    }
    
       
    handleAgentFeature(outputData, requestStartTime) {
        
        /* If feature being watched is 'Agent', this method will handle agent 
       specific tracing logging capabilities*/
       
        this.sessionId = null;
        let prevTraceTime = null;
        return outputData.map(data => {
            if (typeof data === 'object' && data !== null && 'trace' in data) {
                const trace = data.trace;
                if ('start_trace_time' in trace) {
                    //Check if 'start_trace_time' is defined correctly
                    if (typeof trace['start_trace_time'] !== 'number'){
                        throw new Error("The key 'start_trace_time' should be present and should be a time.time() object.");
                    }
                    //Calculate the latency between traces
                    if (prevTraceTime === null) {
                        trace['latency'] = trace['start_trace_time'] - requestStartTime;
                    } else {
                        trace['latency'] = trace['start_trace_time'] - prevTraceTime;
                    }

                    prevTraceTime = trace['start_trace_time'];
                    trace['step_number'] = this.stepCounter;
                    this.stepCounter += 1;
                    data['trace'] = trace;
                }
            } else if (Array.isArray(data)) {
                return data.map(item => {
                    //Check if 'start_trace_time' is defined correctly
                    if (typeof item === 'object' && item !== null && 'start_trace_time' in item) {
                        if (typeof item['start_trace_time'] !== 'number') {
                            throw new Error("The key 'start_trace_time' should be present and should be a time.time() object.");
                        }
                        //Calculate the latency between traces
                        if (prevTraceTime === null) {
                            item['latency'] = item['start_trace_time'] - requestStartTime;
                        } else {
                            item['latency'] = item['start_trace_time'] - prevTraceTime;
                        }

                        prevTraceTime = item['start_trace_time'];
                        item['step_number'] = this.stepCounter;
                        this.stepCounter += 1;
                    } else if (typeof item === 'object' && item !== null && 'trace' in item) {
                        const trace = item['trace'];
                        if ('start_trace_time' in trace) {
                            //Check if 'start_trace_time' is defined correctly
                            if (typeof trace['start_trace_time'] !== 'number') {
                                throw new Error("The key 'start_trace_time' should be present and should be a time.time() object.");
                            }
                            //Calculate the latency between traces
                            if (prevTraceTime === null) {
                                trace['latency'] = trace['start_trace_time'] - requestStartTime;
                            } else {
                                trace['latency'] = trace['start_trace_time'] - prevTraceTime;
                            }

                            prevTraceTime = trace['start_trace_time'];
                            trace['step_number'] = this.stepCounter;
                            this.stepCounter += 1;
                            //Update the 'trace' dictionary in the original item
                            item['trace'] = trace;
                        }
                    }
                    return item;
                });
            }
            return data;
        });
    }

       
    watch(inputs={}) {
        
        /* This method provides the decorator functionality that will be used to log 
       data within the watched function */
        
        var { captureInput = true, captureOutput = true, callType = 'LLM' } = inputs;
        let runId;
        return (func) => {
            return async (...args) => {
                
                //For Latency Calculation
                this.requestStartTime = performance.now();
                
                //Get the function name
                const functionName = func.name;
                
                //Capture input if requested
                const inputData = captureInput ? args : null;
                
                let inputLog = null
                if (inputData) {
                    inputLog = inputData[0];
                }
                
                //Generate observation_id
                const observationId = uuidv4();
                const obsTimestamp = new Date().toISOString();
                
                //Get the start time
                const startTime = performance.now();
                
            
                var outputData;
                let result;
                
                //Calls the function to be executed, the decorated function
                try {
                result = await func(...args);
                //console.log('results',result)
                //Capture output if requested
                outputData = captureOutput ? result : null;
                } 
                catch (error) {
                console.error('Error:', error);
                }
                
                // Get the end time
                const endTime = performance.now();
                
                //Calculate the duration (ms)
                const duration = endTime - startTime;
                
                //Begin Logging Time:
                const loggingStartTime = performance.now();
                
                let runId;

                if (this.featureName !== "Agent") {
                    this.stepCounter = 0;
                }
                
                //Handle the 'Agent' feature case
                if (this.featureName === "Agent") {
                    if (outputData !== null) {
                        outputData = this.handleAgentFeature(outputData, this.requestStartTime);
                        runId = this.extractSessionId(outputData);
                    }
                    else {
                        runId = this.extractSessionId(inputLog)
                    }
                } else {
                    //Extract the session ID from the log or generate a new one
                    runId = this.extractSessionId(inputLog);
                }
                
                //Prepare the metadata
                const metadata = {
                    experiment_id: this.experimentId,
                    run_id: runId,
                    observation_id: observationId,
                    obs_timestamp: obsTimestamp,
                    start_time: new Date(startTime).toISOString(),
                    end_time: new Date(endTime).toISOString(),
                    duration,
                    input_log: inputLog,
                    output_log: outputData,
                    call_type: callType || this.defaultCallType,
                };
                
                //Update the metadata with additional_metadata if provided
                const additionalMetadata = args[args.length - 1]?.additional_metadata || {};
                Object.assign(metadata, additionalMetadata);

                const inputDataVar = args[args.length - 1]?.user_prompt || {};
                Object.assign(metadata, inputDataVar);
                
                //Get the end time
                const loggingEndTime = performance.now();
                const loggingDuration = loggingEndTime - loggingStartTime;
                
                // logging duration is in milliseconds
                metadata['logging_duration'] = loggingDuration;
                
                //Send the metadata to Amazon Kinesis Data Firehose or return it locally for testing:
                if (this.deliveryStreamName ==='local') {
                    if (this.feedbackVariables){
                        console.log("Logs in local mode-with feedback:");
                        return [result,JSON.stringify(metadata),runId,observationId];
                    }
                    else{
                        console.log("Logs in local mode-without feedback:");
                        return [result,JSON.stringify(metadata)];
                    }
                } 
                
                //log to firehose
                else {
                    const firehoseResponse = this.firehoseClient.putRecord({
                        DeliveryStreamName: this.deliveryStreamName,
                        Record: {
                            Data: JSON.stringify(metadata)
                        }
                    }).promise();
                    if (this.feedbackVariables){
                        console.log("Logs in S3-with feedback:");
                        return [result,runId,observationId];
                    }
                    else{
                        console.log("Logs in S3-without feedback:");
                        return result;
                    }
                }
            };
        };
    }
}

export default BedrockLogs;